exercises/tom_clayton/ex02_oo_basics/src/Shape.h

```diff
@@ -0,0 +1,35 @@
+#ifndef SHAPE_H_
+#define SHAPE_H_
+
+#include <string>
+
+class Shape
+{
+  protected: 
+    double m_perimeter;
+    double m_area;
+    std::string m_type;
+    int m_sides;
+    
+  public:
+    Shape();
+
+    std::string shapeToString();
+
+    //Virtual functions
+    virtual double calcPerimeter(double h, double b);

```

Comment:
You can make these pure virtual which means that may class that inherits from this class has to override this function. If you do this then you can include the formula for calculating perimeter/area in the overridden functions rather than passing it into shape.

To override the functions you just declare the function in the subclass as you have declared `calcPerimeter` in Triangle. You will need to do it for each shape